packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -24,15 +24,28 @@ TextAffinity? _toTextAffinity(String? affinity) {
   return null;
 }
 
-/// Replaces a range of text in the original string with the text given in the
-/// replacement string.
+// Replaces a range of text in the original string with the text given in the
+// replacement string.
 String _replace(String originalText, String replacementText, int start, int end) {
+  assert(_textRangeIsValid(TextRange(start: start, end: end), originalText));
   final String textStart = originalText.substring(0, start);
   final String textEnd = originalText.substring(end, originalText.length);
   final String newText = textStart + replacementText + textEnd;
   return newText;
 }
 
+// Verify that the given range is within the text.
+bool _textRangeIsValid(TextRange range, String text) {
+  if (range.start == -1 && range.end == -1) {
+    return true;
+  }
+  assert(range.start >= 0 && range.start <= text.length,
+      'Range start ${range.start} is out of text of length ${text.length}');
```

Comment:
nit: maybe print the entire range here to give more context? The end could be OOB too.

Reply:
nit: looking at the callers of this function, it's a bit difficult to tell what kind of TextRange is OOB, the only thing that can help developers to trace that down is the line number in the stacktrace.

Reply:
I updated this a bit, to make it more obvious what range was failing (delta/selection/composing).